cli/command/service/opts.go

```diff
@@ -335,32 +335,56 @@ type credentialSpecOpt struct {
 	source string
 }
 
+type credentialSpecType string
+
+const (
+	credentialSpecConfig   credentialSpecType = "config"
+	credentialSpecFile     credentialSpecType = "file"
+	credentialSpecRegistry credentialSpecType = "registry"
+)
+
 func (c *credentialSpecOpt) Set(value string) error {
+	// TODO(thaJeztah): should c.source always be set, even if we may error further down?
 	c.source = value
```

Comment:
This was existing behavior, but felt a bit "odd" to set the value regardless if it was valid.

Reply:
Looking at go stdlib flags; the `Set` validates before setting the value (`c.source` is what's returned by `c.String()`; https://github.com/golang/go/blob/7a1679d7ae32dd8a01bd355413ee77ba517f5f43/src/flag/flag.go#L221-L233

spf13/pflags is a bit more ambiguous; it sets the value regardless (although the function used may return a default value); https://github.com/docker/cli/blob/6686ada6a4cf4e872af4d9bf3f61c387be934d87/vendor/github.com/spf13/pflag/uint64.go#L8-L23